app/assets/javascripts/discourse/app/controllers/second-factor-auth.js

```diff
@@ -175,43 +175,37 @@ export default class SecondFactorAuthController extends Controller {
     this.set("messageIsError", false);
   }
 
-  verifySecondFactor(data) {
+  async verifySecondFactor(data) {
     this.set("isLoading", true);
 
-    return ajax("/session/2fa", {
-      type: "POST",
-      data: {
-        ...data,
-        second_factor_method: this.shownSecondFactorMethod,
-        nonce: this.nonce,
-      },
-    })
-      .then((response) => {
-        this.displaySuccess(i18n("second_factor_auth.redirect_after_success"));
-
-        return ajax(response.callback_path, {
-          type: response.callback_method,
-          data: {
-            second_factor_nonce: this.nonce,
-            ...response.callback_params,
-          },
-        })
-          .then((callbackResponse) => {
-            const redirectUrl =
-              callbackResponse.redirect_url || response.redirect_url;
-            return DiscourseURL.routeTo(redirectUrl);
-          })
-          .catch((error) => this.displayError(extractError(error)))
-          .finally(() => {
-            this.set("isLoading", false);
-          });
-      })
-      .catch((error) => {
-        this.displayError(extractError(error));
-      })
-      .finally(() => {
-        this.set("isLoading", false);
+    try {
+      const response = await ajax("/session/2fa", {
+        type: "POST",
+        data: {
+          ...data,
+          second_factor_method: this.shownSecondFactorMethod,
+          nonce: this.nonce,
+        },
+      });
+
+      const callbackResponse = await ajax(response.callback_path, {
+        type: response.callback_method,
+        data: {
+          second_factor_nonce: this.nonce,
+          ...response.callback_params,
+        },
       });
+
+      this.displaySuccess(i18n("second_factor_auth.redirect_after_success"));
+
+      DiscourseURL.routeTo(
+        callbackResponse.redirect_url || response.redirect_url
+      );
+    } catch (error) {
+      this.displayError(extractError(error));
+    } finally {
+      this.set("isLoading", false);
+    }
```

Comment:
Refactoring to `async..await` since the nested promise chain makes it hard to understand.